### PR TITLE
Explicit api for all modules (and small changes on the way)

### PR DESCRIPTION
**Pull request by Lukellmann**
Because this PR has a lot of changed lines, I will highlight those changes that are more than just adding `public`, specifying types explicitly and formatting, in review comments.

Notable other changes:
- grouped routes by resource in `rest.route.Route`
- corrected Guild Scheduled events routes (had `events` instead of `scheduled-events` in their paths), see https://discord.com/developers/docs/resources/guild-scheduled-event
- removed generated / unused `companion objects`
- `GuildCreateBuilder.snowflakeGenerator` is now of type `Iterator<Snowflake>`, no longer `Iterator<ULong>`
- public non-companion `(De)SerializationStrategy` object where no full serializer exists

### CODE REVIEW

File: rest/src/main/kotlin/route/Route.kt [hunk 1]
```diff
@@ -73,789 +40,837 @@ internal class NullAwareMapper<T>(val strategy: DeserializationStrategy<T>) : Re
...
-    val mapper: ResponseMapper<T>
+internal object NoStrategy : DeserializationStrategy<Unit> {
+    @OptIn(InternalSerializationApi::class, ExperimentalSerializationApi::class)
+    override val descriptor: SerialDescriptor
+        get() = buildSerialDescriptor("NoStrategy", StructureKind.OBJECT) {}
+
+    override fun deserialize(decoder: Decoder) {}
+}
+
+public sealed class Route<T>(
```
Comment (Lukellmann) [inline rest/src/main/kotlin/route/Route.kt:51]: grouped routes similar to docs (didn't add, remove or change any)

File: common/src/main/kotlin/entity/DiscordGuild.kt [hunk 2]
```diff
@@ -294,18 +294,14 @@ sealed class GuildFeature(val value: String) {
     }
 }
 
-@Serializable(with = SystemChannelFlags.Companion::class)
-class SystemChannelFlags constructor(val code: Int) {
+@Serializable(with = SystemChannelFlags.Serializer::class)
+public data class SystemChannelFlags(val code: Int) {
```
Comment (Lukellmann) [inline common/src/main/kotlin/entity/DiscordGuild.kt:298]: `data class` like other Flags classes (see https://github.com/kordlib/kord/pull/471#discussion_r776903782)

File: rest/src/main/kotlin/route/CdnUrl.kt [hunk 3]
```diff
@@ -23,5 +23,8 @@ class CdnUrl(private val rawAssetUri: String) {
         return "CdnUrl(rawAssetUri=$rawAssetUri)"
     }
 
-    data class UrlFormatBuilder(var format: Image.Format = Image.Format.WEBP, var size: Image.Size? = null)
-}
\ No newline at end of file
+    public class UrlFormatBuilder {
+        public var format: Image.Format = Image.Format.WEBP
+        public var size: Image.Size? = null
+    }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/route/CdnUrl.kt]: no `data class` for builders

File: voice/src/main/kotlin/gateway/VoiceEvent.kt [hunk 4]
```diff
@@ -127,7 +127,7 @@ public sealed class Close : VoiceEvent() {
      *
      * @param recoverable true if the gateway will automatically try to reconnect.
      */
-    public class DiscordClose(public val closeCode: VoiceGatewayCloseCode, public val recoverable: Boolean) : Close()
+    public data class DiscordClose(val closeCode: VoiceGatewayCloseCode, val recoverable: Boolean) : Close()
```
Comment (Lukellmann) [inline voice/src/main/kotlin/gateway/VoiceEvent.kt:130]: `data class` like all other voice `Event` classes

File: rest/src/main/kotlin/route/CdnUrl.kt [hunk 5]
```diff
@@ -23,5 +23,8 @@ class CdnUrl(private val rawAssetUri: String) {
...
+}
```
Comment (Lukellmann) [inline rest/src/main/kotlin/route/CdnUrl.kt]: no `data class` for builders

**Comment by Lukellmann** [comment]
I don't think it was a good idea to pack all of these changes into a single PR. I will therefore close this now and open several smaller PRs later.
I hope this will not lead to more work for any of you.

**Comment by HopeBaron** [comment]
I didn't plan to close this, rather it's fine to merge. that is expected from such a PR

File: common/src/main/kotlin/entity/Interactions.kt [hunk 6]
```diff
@@ -4,40 +4,15 @@ import dev.kord.common.annotation.KordExperimental
...
-import kotlinx.serialization.json.JsonPrimitive
-import kotlinx.serialization.json.double
-import kotlinx.serialization.json.doubleOrNull
-import kotlinx.serialization.json.jsonObject
-import kotlinx.serialization.json.jsonPrimitive
-import kotlinx.serialization.json.long
-import kotlinx.serialization.json.longOrNull
-import mu.KotlinLogging
-
-val kordLogger = KotlinLogging.logger { }
```
Comment (Lukellmann) [inline common/src/main/kotlin/entity/Interactions.kt:37]: Unused and seems to be out of place in package ` dev.kord.common.entity`.

File: rest/src/main/kotlin/json/response/DiscordErrorResponse.kt [hunk 7]
```diff
@@ -1,20 +1,14 @@
...
-import kotlinx.serialization.encoding.Encoder
-import kotlinx.serialization.json.*
+import kotlinx.serialization.json.JsonElement
 
 /**
  * Represents a [Discord error response](https://discord.com/developers/docs/topics/opcodes-and-status-codes#json).
  */
 @Serializable
-class DiscordErrorResponse(
+public data class DiscordErrorResponse(
```
Comment (Lukellmann) [inline rest/src/main/kotlin/json/response/DiscordErrorResponse.kt:11]: `data class` seems appropriate

File: rest/src/main/kotlin/service/ChannelService.kt [hunk 8]
```diff
@@ -17,139 +17,158 @@ import dev.kord.rest.route.Route
...
-    suspend fun deleteChannel(channelId: Snowflake, reason: String? = null) = call(Route.ChannelDelete) {
-        keys[Route.ChannelId] = channelId
-        auditLogReason(reason)
-    }
+    public suspend fun bulkDelete(channelId: Snowflake, messages: BulkDeleteRequest, reason: String?): Unit =
+        call(Route.BulkMessageDeletePost) {
+            keys[Route.ChannelId] = channelId
+            body(BulkDeleteRequest.serializer(), messages)
+            auditLogReason(reason)
+        }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/service/ChannelService.kt:135]: was missing for `ChannelService.bulkDelete()`

File: rest/src/main/kotlin/service/ChannelService.kt [hunk 9]
```diff
@@ -197,96 +219,95 @@ class ChannelService(requestHandler: RequestHandler) : RestService(requestHandle
...
 
+    public suspend fun putChannel(
+        channelId: Snowflake,
+        channel: ChannelModifyPutRequest,
+        reason: String? = null,
+    ): DiscordChannel = call(Route.ChannelPut) {
+        keys[Route.ChannelId] = channelId
+        body(ChannelModifyPutRequest.serializer(), channel)
+        auditLogReason(reason)
+    }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/service/ChannelService.kt:264]: was `reason?.let { header("X-Audit-Log-Reason", reason) }` before

File: rest/src/main/kotlin/service/ChannelService.kt [hunk 10]
```diff
@@ -197,96 +219,95 @@ class ChannelService(requestHandler: RequestHandler) : RestService(requestHandle
...
-        }
+    public suspend fun patchThread(
+        threadId: Snowflake,
+        thread: ChannelModifyPatchRequest,
+        reason: String? = null,
+    ): DiscordChannel = call(Route.ChannelPatch) {
+        keys[Route.ChannelId] = threadId
+        body(ChannelModifyPatchRequest.serializer(), thread)
+        auditLogReason(reason)
+    }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/service/ChannelService.kt:284]: was `reason?.let { header("X-Audit-Log-Reason", reason) }` before

File: rest/src/main/kotlin/service/ChannelService.kt [hunk 11]
```diff
@@ -197,96 +219,95 @@ class ChannelService(requestHandler: RequestHandler) : RestService(requestHandle
...
-            body(StartThreadRequest.serializer(), request)
-            reason?.let { header("X-Audit-Log-Reason", reason) }
-        }
+        reason: String? = null,
+    ): DiscordChannel = call(Route.StartPublicThreadWithMessagePost) {
+        keys[Route.ChannelId] = channelId
+        keys[Route.MessageId] = messageId
+        body(StartThreadRequest.serializer(), request)
+        auditLogReason(reason)
     }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/service/ChannelService.kt:308]: `reason?.let { header("X-Audit-Log-Reason", reason) }` -> `auditLogReason(reason)`

File: rest/src/main/kotlin/service/ChannelService.kt [hunk 12]
```diff
@@ -298,19 +319,17 @@ class ChannelService(requestHandler: RequestHandler) : RestService(requestHandle
...
-            keys[Route.ChannelId] = channelId
-            body(StartThreadRequest.serializer(), request)
-            reason?.let { header("X-Audit-Log-Reason", reason) }
-        }
+        reason: String? = null,
+    ): DiscordChannel = call(Route.StartThreadPost) {
+        keys[Route.ChannelId] = channelId
+        body(StartThreadRequest.serializer(), request)
+        auditLogReason(reason)
     }
```
Comment (Lukellmann) [inline rest/src/main/kotlin/service/ChannelService.kt:330]: `reason?.let { header("X-Audit-Log-Reason", reason) }` -> `auditLogReason(reason)`

**Comment by Lukellmann** [comment]
After some work on this PR (like moving some changes to https://github.com/kordlib/kord/pull/482) I would say it has reached a point where it can be reviewed.

My recommendations for viewing the diffs:
- some files in `/rest/src/main/kotlin/service` are best viewed with a split diff (formatting changes make the unified view rather confusing)
- a lot of lines changed for `/rest/src/main/kotlin/route/Route.kt` -> view the diffs for that file **_commitwise_**

File: rest/src/main/kotlin/service/InteractionService.kt [hunk 13]
```diff
@@ -600,12 +565,10 @@ class InteractionService(requestHandler: RequestHandler) : RestService(requestHa
...
 
-
-    suspend inline fun modifyInteractionResponse(
+    public suspend inline fun modifyInteractionResponse(
         applicationId: Snowflake,
         interactionToken: String,
-        ephemeral: Boolean = false,
-        builder: InteractionResponseModifyBuilder.() -> Unit
+        builder: InteractionResponseModifyBuilder.() -> Unit,
     ): DiscordMessage {
```
Comment (Lukellmann) [inline rest/src/main/kotlin/service/InteractionService.kt:572]: `ephemeral` option does not exist for this endpoint (param was unused)

**Pull request merged**